### PR TITLE
DHT rework, part 1: Low-level infrastructure

### PR DESCRIPTION
**Pull request by Renelvon**
### Motivation
The DHT and the rest of OpenBazaar codebase are tangled, and there is no real DHT API. This PR is an attempt to address some of the technical debt.

### Changelog
* Create a `/dht` folder for all things DHT-ish.
* Fix some violations of the Kademlia protocol.
* Recreate a `Contact` class, according to the Kademlia specification. This class is pretty small right now, but there is ample room for growth.
* Move a lot of code out of `RoutingTable` and into `KBucket`, following the [Tell, Don't Ask](http://martinfowler.com/bliki/TellDontAsk.html) principle. Move some state-agnostic code to a separate `util` module.
* Make the API a little less Pythonic; provide separate methods for separate types (i.e. `Contact`s and `str`ings). Make the methods in question much easier to read and test, at the expense of some code duplication.
* Have `KBucket` and `RoutingTable` implement the [collections.Sequence](https://docs.python.org/2/library/collections.html#collections-abstract-base-classes) protocol. Provide read-only access to the underlying attributes without explicit getter methods, and make the objects usable in iterator context (i.e. in a `for` loop). Improve encapsulation by making a lot of attributes and methods private (names preceded by `_`). 
* Explicitly document all API methods, following the [Google Code guide](https://google-styleguide.googlecode.com/svn/trunk/pyguide.html#Comments).
* Test everything, thoroughly. Subsequent changes should also be accompanied by the appropriate tests.

### Notes
* This infrastructure is not a working DHT, and is not used by the node yet. So, there is no way to test it functionally. Please review the API and the tests in `tests/dht`.
* The `/dht` folder is not included in the test coverage right now, because it would be misleading. It will be, once the new DHT is fully functional and integrated.

### CODE REVIEW

**Comment by SamPatt** [comment]
+1 Great to see you starting in on the DHT.

**Comment by hoffmabc** [comment]
I'm reviewing but is this ready to merge if it's good or is this WIP?

**Comment by Renelvon** [comment]
@hoffmabc This part is complete. However, it is not working yet, as it lacks a proper DHT built on top of it. Once the new DHT is complete and integrated, we will have the client use this library. Does that make sense?

**Comment by hoffmabc** [comment]
That's what I thought. I was just checking.

File: dht/contact.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import collections
+
+
+class Contact(collections.Hashable):
+    """
```
Comment (gubatron) [inline dht/contact.py]: YES!
yes-pickard

File: dht/kbucket.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        Move contacts from the cache to the main list, until the
+        cache is exhausted or the main list is full.
+        """
+        move_count = min(
+            len(self._replacement_cache),
+            constants.K - len(self._contacts)
+        )
+
+        for _ in range(move_count):
+            self.add_contact(self._replacement_cache.pop())
```
Comment (gubatron) [inline dht/kbucket.py]: great job @Renelvon

**Comment by Renelvon** [comment]
Oops, I didn't fix the constants in time.

**Comment by hoffmabc** [comment]
You can just tweak them in the next update. It's not a big deal as this code is dormant for now @Renelvon.

**Pull request merged**